### ISSUE TITLE
Different cost values if the start and end points are switched

### ISSUE THREAD

**Issue by QasimMuhammad** [scikit-image/scikit-image#3380]
Just found out while using route_through_array gives different cost when start and end points are switched and hence result in different paths. 

   indices, weight = route_through_array(costArray, (startIndexY, startIndexX), (stopIndexY, stopIndexX),geometric=True, fully_connected=True)

The costArray consist of 1 and 255 values. The path is made through pixel with value one.

cost value = 641.3
 
and when start and end points are switched the cost become

cost value = 741.2

I understand some difference may be expected but the difference here is larger than what can be tolerated. 

Any suggestions?

**Comment by jni** [comment]
Thanks for the report @QasimMuhammad! This seems like a bug to me. @zpincus would you mind looking at this?

@QasimMuhammad in the meantime, if you could provide us with an example image and start and end values to reproduce this problem, it would be extremely useful! Without it, we are taking shots in the dark.

**Comment by zpincus** [comment]
Thanks @QasimMuhammad! This is almost certainly a bug.

I'll certainly be happy to take a look if you can get a minimal example together that I could dig into.

**Comment by jni** [comment]
@QasimMuhammad indeed, a naive test example doesn't reproduce your error:

```python
import numpy as np
from skimage import graph

np.random.seed(42)
costs = np.where(np.random.random((512, 512)) > 0.5, 1, 255)
start, end = (0, 128), (511, 386)

print(graph.route_through_array(costs, start, end, geometric=True, fully_connected=True)[1])
print(graph.route_through_array(costs, end, start, geometric=True, fully_connected=True)[1])
```

results in:

```
1193.6072144624395
1193.6072144624395
```

So there is something strange going on with your specific data/use-case, and we would need your data with a minimal test script to hunt down the error.

**Comment by QasimMuhammad** [comment]
@jni, I don't want to share my whole code. In order to send minimal part of it, I need to work on it a bit while I am busy with other things. Give me couple of days and I will share. Thanks for your support

**Comment by jni** [comment]
@QasimMuhammad absolutely, and we don't want your whole code either, as that makes it harder to find the exact source of the bug. =P Just wanted to make it clear that the ball is in your court!

**Comment by RafaelCarlquist** [comment]
Hello everyone, I was wondering if someone else took a look at this question. I think the problem happens when the end/start point is not an allowed point (with negative values, for example). In this case, changing the order of what is the end and what is the start will produce a different result. 

Maybe it is related with #1912

I wrote a small example:

```
# libraries
import numpy as np
import skimage.graph as skimg 

# matrix cost
cost = np.array([[-1,1,1,1,1],
                 [1,1,1,1,1],
                 [1,1,1,1,1]])

# generating instance
class_mcp = skimg.MCP(cost, fully_connected = False)

# computing cost
cum_cost , aux = class_mcp.find_costs(starts = [[0,0]], ends = [[2,4]], find_all_ends = True)

# printing
cum_cost
```

Here, the fisrt cell is -1, which is weird. Now, if we change the order

```
cum_cost , aux = class_mcp.find_costs(starts = [[2,4]], ends = [[0,0]], find_all_ends = True)

# printing
cum_cost
```

The first cell has the correct value of infinity.

Thank you all,

Rafael

**Comment by jni** [comment]
@RafaelCarlquist thanks for reviving this! If that is indeed the issue, it might be quite easy to fix!
